persistence/trees/trees.go

```diff
@@ -251,6 +272,66 @@ func (t *treeStore) getStateHash() string {
 	return hexHash
 }
 
+////////////////////////////////
+// AtomicStore Implementation //
+////////////////////////////////
+
+// Savepoint generates a new savepoint for the tree store and saves it internally.
+func (t *treeStore) Savepoint() error {
+	w, err := t.save()
+	if err != nil {
+		return err
+	}
+	t.Prev = w
+	return nil
+}
+
+// Rollback intentionally can't return an error because at this point we're out of tricks
+// to recover from problems.
+func (t *treeStore) Rollback() {
+	if t.Prev != nil {
+		t.merkleTrees = t.Prev.MerkleTrees
+		t.rootTree = t.Prev.RootTree
+		return
+	}
+	t.logger.Fatal().Msgf("rollback called without valid savepoint - this is a bug - treeStore shutting down: %+v", t)
+}
+
+// save commits any pending changes to the trees and serializes the treeStore to a WorldState object
+// with new readers and writers.
+// OPTIMIZE: Consider saving only the root hash of each tree and the tree directory here and then
+// load the trees up in Rollback instead of setting them up here.
+func (t *treeStore) save() (*Worldstate, error) {
+	if err := t.Commit(); err != nil {
+		return nil, err
+	}
+
+	w := &Worldstate{
+		TreeStoreDir: t.treeStoreDir,
+		MerkleTrees:  map[string]*stateTree{},
+	}
+
+	for treeName := range t.merkleTrees {
```

Comment:
Can you add some inline comments about how this creates a whole new copy of the trees?